packages/core/src/api/serialization/serialization.test.ts

```diff
@@ -0,0 +1,296 @@
+import { afterEach, beforeEach, describe, expect, it } from "vitest";
+import { BlockNoteEditor } from "../../BlockNoteEditor";
+import { Editor } from "@tiptap/core";
+import { customBlockSerializer } from "./serialization";
+import { uploadToTmpFilesDotOrg_DEV_ONLY } from "../../extensions/Blocks/nodes/BlockContent/ImageBlockContent/uploadToTmpFilesDotOrg_DEV_ONLY";
+import { createBlockSpec } from "../../extensions/Blocks/api/block";
+import {
+  imagePropSchema,
+  renderImage,
+} from "../../extensions/Blocks/nodes/BlockContent/ImageBlockContent/ImageBlockContent";
+import { defaultBlockSchema } from "../../extensions/Blocks/api/defaultBlocks";
+import { cleanHTML } from "../formatConversions/formatConversions";
+import { DOMSerializer, Fragment, Node } from "prosemirror-model";
+
+// This is a modified version of the default image block that does not implement
+// a `serialize` function. It's used to test if the custom serializer by default
+// serializes custom blocks using their `render` function.
+const SimpleImage = createBlockSpec({
+  type: "simpleImage",
+  propSchema: imagePropSchema,
+  containsInlineContent: false,
+  render: renderImage as any,
+});
+
+const customSchema = {
+  ...defaultBlockSchema,
+  simpleImage: SimpleImage,
+};
+
+let editor: BlockNoteEditor<typeof customSchema>;
+let tt: Editor;
+
+beforeEach(() => {
+  (window as Window & { __TEST_OPTIONS?: any }).__TEST_OPTIONS = {};
+
+  editor = new BlockNoteEditor({
+    blockSchema: customSchema,
+    uploadFile: uploadToTmpFilesDotOrg_DEV_ONLY,
+  });
+  tt = editor._tiptapEditor;
+});
+
+afterEach(() => {
+  tt.destroy();
+  editor = undefined as any;
+  tt = undefined as any;
+
+  delete (window as Window & { __TEST_OPTIONS?: any }).__TEST_OPTIONS;
+});
+
+// Tests `serializeNode`, `serializeFragment`, and `cleanHTML` methods.
+async function serializeAndCompareSnapshots(
+  serializer: DOMSerializer,
+  blockGroup: Node
+) {
+  const serializeNodeHTML = (
+    serializer.serializeNode(blockGroup) as HTMLElement
+  ).outerHTML;
+  const serializeFragmentHTML = (
+    serializer.serializeFragment(Fragment.from(blockGroup)) as HTMLElement
+  ).firstElementChild!.outerHTML;
+
+  expect(serializeNodeHTML).toMatchSnapshot();
+  expect(serializeFragmentHTML).toMatchSnapshot();
+
+  const serializeNodeCleanHTML = await cleanHTML(serializeNodeHTML);
+  const serializeFragmentCleanHTML = await cleanHTML(serializeFragmentHTML);
+
+  expect(serializeNodeCleanHTML).toMatchSnapshot();
+  expect(serializeFragmentCleanHTML).toMatchSnapshot();
+}
+
+describe("Convert paragraphs to structured HTML", () => {
+  it("Convert paragraph to structured HTML", async () => {
+    const serializer = customBlockSerializer(tt.schema, editor);
+
+    const text = tt.schema.text("Paragraph");
```

Comment:
Is there a reason we use the pm syntax to instantiate nodes / a document, instead of the blocknote API? Better to "eat your own dogfood"?

Reply:
2 reasons:
1. The BlockNote API doesn't let you create blocks outside the editor, i.e. when creating a block, you have to insert it somewhere into the editor. We already have unit tests for block insertion, and I think since these ones are meant to test serialization specifically, it's better that they don't fail if something non-serialization related goes wrong.